### DISCUSSION TITLE
How to use Prisma 'set' with explicit many-to-many relation?

### DISCUSSION THREAD

**Discussion by ngoc199** [prisma/prisma#10048] [Q&A]
As in this [document](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#set) of Prisma, `set` can be used to override the value of a relation.

```ts
const user = await prisma.user.update({
  where: { email: 'alice@prisma.io' },
  data: {
    posts: {
      set: [{ id: 32 }, { id: 42 }],
    },
  },
})
```

But when I tried it with explicit many-to-many relation, it does not work.

```prisma
model Product {
  id        String   @id @default(cuid())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  name        String
  description String?
  price       Decimal
  sku         String  @unique
  published   Boolean @default(false)

  tags ProductTag[]
}

model Tag {
  id   Int    @id @default(autoincrement())
  name String

  products ProductTag[]
}

model ProductTag {
  productId String
  tagId     Int

  createdAt DateTime @default(now())

  product Product @relation(fields: [productId], references: [id])
  tag     Tag     @relation(fields: [tagId], references: [id])

  @@id([productId, tagId])
}

```

My code to update `Product`

```ts
update(id: string, updateProductDto: UpdateProductDto) {
    const tags = updateProductDto.tags.map((tag) => ({
      productId_tagId: {
        productId: id,
        tagId: tag.id,
      },
    }));

    console.log(JSON.stringify(tags));

    return this.prisma.product.update({
      where: { id: id },
      data: {
        ...updateProductDto,
        tags: {
          set: [...tags],
        },
      },
    });
  }
```
I want to update the product's tag with the product's information.

How could I implement this correctly?

**Comment by silkyland** [comment]
I got the same problem

**Comment by austincrim** [comment]
Hey @ngoc199 👋, thanks for the question.

When you say that this operation doesn't work, are you getting a specific error or is the change just not taking effect? If it's an error, could you provide me with the message?

Also, would you mind sharing some sample data for your 3 models so I can accurately reproduce the issue?

Thanks!

**Comment by goranefbl** [comment]
I got the same problem, but since my join table has unique key (due to sql_require_primary_key set, and perf. reasons), I get this error: 

```
Argument data.jobs_levels.set of type jobs_levelsWhereUniqueInput needs exactly one argument, but you provided levelId and jobId. Please choose one. Available args: 
type jobs_levelsWhereUniqueInput {
  id?: Int
}
```

when I try to use:

`jobs_levels: { set: [ {levelId: 2, jobId: 18}]}`

**Comment by goranefbl** [comment]
Without unique key, I get this error: 


```

Argument data.jobs_levels.set of type jobs_levelsWhereUniqueInput needs exactly one argument, but you provided levelId and jobId. Please choose one. Available args: 
type jobs_levelsWhereUniqueInput {
  levelId_jobId?: jobs_levelsLevelIdJobIdCompoundUniqueInput
}
Unknown arg `levelId` in data.jobs_levels.set.0.levelId for type jobs_levelsWhereUniqueInput. Did you mean `select`? Available args:
type jobs_levelsWhereUniqueInput {
  levelId_jobId?: jobs_levelsLevelIdJobIdCompoundUniqueInput
}
Unknown arg `jobId` in data.jobs_levels.set.0.jobId for type jobs_levelsWhereUniqueInput. Available args:

type jobs_levelsWhereUniqueInput {
  levelId_jobId?: jobs_levelsLevelIdJobIdCompoundUniqueInput
}
```

**Reply by austincrim** [reply]
Hey @goranefbl, could you share your Prisma schema and relevant code?

**Comment by AustinGil** [comment]
I'm in this same boat. I have things setup with implicit relations fine, but with explicit relations I keep getting the same errors.

Prisma schema
```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model Person {
  id      String      @id
  name    String
  pets   PersonOnPet[]
}

model Pet {
  id      String      @id
  name    String
  type    String
  birthday DateTime   @default(now())
  owners   PersonOnPet[]
}

model PersonOnPet {
  person      Person     @relation(fields: [personId], references: [id])
  personId    String // relation scalar field (used in the `@relation` attribute above)
  pet         Pet        @relation(fields: [petId], references: [id])
  petId       String // relation scalar field (used in the `@relation` attribute above)

  @@id([personId, petId])
}
```

JS Query:
```js
db.pet.update({
      where: { id: petId },
      data: {
        owners: {
          connect: [
            {
              personId_petId: {
                personId: ownerId,
                petId,
              },
            },
          ],
        },
      },
      include: {
        owners: true,
      },
    })
```

The error:
`The required connected records were not found. Expected 1 records to be connected after connect operation on one-to-many relation 'PersonOnPetToPet', found 0.`

Which sounds like one of the entries does not exist in the DB, but I've confirmed that it does. Anyone else running into this error?

**Comment by steven-tey** [comment]
Related: https://github.com/prisma/prisma/issues/12701

> Your problems stem from a misunderstanding of set: This doesn't delete anything, this disconnects the currently connected relation (the trackId: 1 and playlistId: 1 record from above) and sets a new set (here, empty). If you disconnect a TrackPlaylist record, it will fail because you're violating referential integrity by doing so.
> 
> What you want are the correct operations to achieve the right things: create for creating explicit m2m links, delete for deleting certain ones, not set, which doesn't do any of that.

**Comment by ahoopes16** [comment]
For anyone else coming across this, here's a solution that worked for me. The only thing to note is that it will wipe out the old relations, but I think that's the behavior that was expected for `set` anyway.

### Solution
Because `set` doesn't create or delete records, we need to do both of these steps. Luckily, you can do both of these steps in a single query. I've adapted your code to show what I do in my code:

```typescript
// BEFORE -- NOT WORKING
update(id: string, updateProductDto: UpdateProductDto) {
    const tags = updateProductDto.tags.map((tag) => ({
      productId_tagId: {
        productId: id,
        tagId: tag.id,
      },
    }));

    return this.prisma.product.update({
      where: { id: id },
      data: {
        ...updateProductDto,
        tags: {
          set: [...tags],
        },
      },
    });
  }

// AFTER -- WORKING
update(id: string, updateProductDto: UpdateProductDto) {
    return this.prisma.product.update({
      where: { id },
      data: {
        ...updateProductDto,
        tags: {
          deleteMany: {},
          create: updateProductDto.tags.map((t) => ({ tag: { connect: { t.id } } }))
        },
      },
    });
  }

```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
